### Brief Overview 

- The goal of this assignment is to practice some important data wrangling functionality commonly required in real-world projects.

- Here we will use two datasets:
  - IRS Statistics of Income (SOI) dataset
  - The Medicaid Data per State  


- The final product here is a table with medication cost per Medicaid enrollee per state. This dataset will allow us to answer such questions as:
  - Which medications account for the bulk of a state's spending   
  - Which drugs are prescribed much more in one state compared to the other states.
etc.


In [1]:
import pandas as pd
import numpy as np


* Load the IRS Statistics of Income (SOI) dataset (tax_data.csv) into a `DataFrame` called `tax_data`. The file is `tax_data.csv` is located in the `data` directory of the assignment folder.

* This dataset was preprocessed but the original one was obtained at the following URL:

        https://www.irs.gov/pub/irs-soi/15zpallagi.csv


In [2]:
## WRITE YOUR CODE HERE 
tax_data = pd.read_csv('data/tax_data.csv')

* Use a `tax_data` method or property to display the first eight (8) rows of the `DataFrame`

In [3]:
## WRITE YOUR CODE HERE 
tax_data.head(8)

STATEFIPS STATE  zipcode  agi_stub        N1     mars1     MARS2     MARS4  \
0        1.0    AL      0.0       1.0  836320.0  481570.0  109790.0  233260.0   
1        1.0    AL      0.0       2.0  494830.0  206630.0  146250.0  129390.0   
2        1.0    AL      0.0       3.0  261250.0   80720.0  139280.0   36130.0   
3        1.0    AL      0.0       4.0  166690.0   28510.0  124650.0   10630.0   
4        1.0    AL      0.0       5.0  212660.0   19520.0  184320.0    4830.0   
5        1.0    AL      0.0       6.0   55360.0    2950.0   49260.0     350.0   
6        1.0    AL  35004.0       1.0    1490.0     970.0     230.0     280.0   
7        1.0    AL  35004.0       2.0    1350.0     630.0     360.0     300.0   

       PREP         N2  ...    N10300     A10300   N85530   A85530   N85300  \
0  455560.0  1356760.0  ...  373410.0   328469.0      0.0      0.0      0.0   
1  275920.0  1010990.0  ...  395880.0   965011.0      0.0      0.0      0.0   
2  155100.0   583910.0  ...  251490.0  1333418.0      0.0      0.0      0.0   
3   99950.0   423990.0  ...  165320.0  1414283.0      0.0      0.0      0.0   
4  126860.0   589490.0  ...  212000.0  3820152.0    420.0    168.0     60.0   
5   41410.0   160530.0  ...   55300.0  6027793.0  22090.0  39519.0  27550.0   
6     700.0     2160.0  ...     690.0      610.0      0.0      0.0      0.0   
7     610.0     2540.0  ...    1140.0     3019.0      0.0      0.0      0.0   

    A85300   N11901    A11901    N11902     A11902  
0      0.0  61920.0   48150.0  732670.0  1933120.0  
1      0.0  73720.0  107304.0  415410.0  1187403.0  
2      0.0  64200.0  139598.0  193030.0   536699.0  
3      0.0  45460.0  128823.0  116440.0   377177.0  
4     31.0  83330.0  421004.0  121570.0   483682.0  
5  95112.0  28590.0  791573.0   15960.0   250289.0  
6      0.0    120.0      94.0    1290.0     2792.0  
7      0.0    210.0     301.0    1130.0     2935.0  

[8 rows x 91 columns]

*  Modify `tax_data` to uppercase all the header name. 
  * Your answer can only use `DataFrame` or `Series` methods or properties
  * Do not hardcode the operation by uppercasing the columns yourself
  *  You can `tax_data.columns`, which returns a `Series` of the column names.
  
* The resulting column name should look as follows:


```
STATEFIPS    STATE    ZIPCODE    AGI_STUB    N1    MARS1    MARS2    MARS4    PREP    N2    ...    
```

This operation is useful for standardizing column names and avoid guessing whether the column header was in upper case, lower case or a mix of both.

In [4]:
tax_data.columns = tax_data.columns.str.upper()

* What is the total number of entries (also called observations) in `tax_data`?

  * Your answer can only use `DataFrame` or `Series` methods or properties


In [5]:
## WRITE YOUR CODE HERE 
tax_data.shape

(166698, 91)

- If `STATEFIPS` is header title of the first column of the `tax_data` `DataFrame`, what is the title of the 32nd column
  - Your answer can only use `DataFrame` or `Series` methods or properties and should use a single python expression



In [6]:
## WRITE YOUR CODE HERE 
tax_data.columns[31]

'N00900'

* If `STATEFIPS` is the the first column, what is the index of the column name `N10300`?

  * Your answer can only use `DataFrame` or `Series` methods or properties


In [7]:
## WRITE YOUR CODE HERE 
tax_data.columns.get_loc('N10300')

81

In [8]:
## WRITE YOUR CODE HERE 
states_zips = tax_data.groupby('STATE')['ZIPCODE'].nunique()
states_zips = states_zips.reset_index()
states_zips = states_zips.sort_values(by='ZIPCODE', ascending=False)
states_zips

STATE  ZIPCODE
43    TX     1619
34    NY     1540
4     CA     1485
38    PA     1368
14    IL     1231
35    OH      998
9     FL      920
22    MI      892
24    MO      890
12    IA      827
45    VA      793
23    MN      792
27    NC      725
48    WI      714
15    IN      674
10    GA      667
17    KY      656
16    KS      598
42    TN      590
1     AL      577
31    NJ      547
36    OK      547
49    WV      512
47    WA      497
2     AR      490
29    NE      486
19    MA      482
18    LA      453
20    MD      405
5     CO      395
40    SC      376
25    MS      370
21    ME      370
37    OR      356
3     AZ      295
28    ND      287
41    SD      287
6     CT      264
46    VT      238
30    NH      232
26    MT      225
13    ID      213
32    NM      209
44    UT      186
33    NV      128
50    WY      109
39    RI       71
11    HI       60
8     DE       57
0     AK       56
7     DC       24

- Identify the position of HI in the list of zip code counts per state (questions directly above)
  - Your answer can only use `DataFrame` or `Series` methods or properties

In [9]:
## WRITE YOUR CODE HERE
states_zips.loc[states_zips['STATE'] == 'HI']

STATE  ZIPCODE
11    HI       60

### Identifying and Removing Ambiguous Zip Codes

- Count the number of entries where ZIPCODE is 0, assign your results to a variable named  `nb_invalid_zip`

In [10]:
## WRITE YOUR CODE HERE 
nb_invalid_zip = tax_data['ZIPCODE'].value_counts()[0.0].tolist()
#type(nb_invalid_zip)
print(nb_invalid_zip)

306


* Run the line below to make sure that `nb_invalid_zip` is an integer (`int`)
  * Note that `assert` will only print an error if `type(nb_invalid_zip)` is not of type `int`
  
* If `nb_invalid_zip` then change your answer above so that the value returned is effectively a number.

In [11]:
assert(type(nb_invalid_zip) == int)

- Remove from `tax_data` all the lines where the zip code is `0` and save resulting `DataFrame` to a variable named `tax_data_valid_zip`
  - Your answer can only use `DataFrame` or `Series` methods or properties


In [12]:
tax_data_valid_zip = tax_data[tax_data['ZIPCODE'] != 0.0]
tax_data_valid_zip.shape

(166392, 91)

* Run the line below to confirm that the operation worked as expected

  * The assertion below is testing that the number of lines with "zip code equal to  0" + number of lines in `tax_data_valid_zip` is equal to the number of lines in the original `DataFrame` `tax_data`
  
  * The assertion below will fail (and print an error message) if the results do not match. If that is the case, please review your code above.

In [13]:
assert((tax_data_valid_zip.shape[0] + nb_invalid_zip) == tax_data.shape[0])

### Identifying and Removing Lines with Missing Values

* How many lines contain at least one missing value `NaN` in the `tax_data_valid_zip` `DataFrame`?
  * Your answer can only use `DataFrame` methods and properties
* Assing the count of `NaN` into a variable called nb_missing_values

In [14]:
## WRITE YOUR CODE HERE
nb_missing_values = tax_data_valid_zip.isnull().any(axis=1).sum()
#tax_data_valid_zip.info()
nb_missing_values

139

* Create a new `DataFrame` containing all the lines from `tax_data_valid_zip`, except lines containing missing values

* Call the new `DataFrame` `tax_data_valid_zip_cleaned

In [15]:
## WRITE YOUR CODE HERE 
tax_data_valid_zip_cleaned = tax_data_valid_zip.dropna(axis=0)
tax_data_valid_zip_cleaned.shape

(166253, 91)

* Run the line below to confirm that the operation worked as expected. The assertion below is testing that:  
`nb_missing_values` + number of lines in `tax_data_valid_zip_cleaned` is equal to the number of lines in `tax_data_valid_zip`
  
* Note that assert will only print an error if the results do not match

In [16]:
assert((tax_data_valid_zip_cleaned.shape[0] + nb_missing_values) == tax_data_valid_zip.shape[0])

### Computing the Percentile Income per Zip Code

* The function `compute_percentile_zipcode` below computes the percentile income per zip code

* By default percentile=0.5,  i.e., the function computes the median

* Read the code and make sure you understand what it does before moving on to the next question


In [17]:
#def compute_percentile_zip(df_zip, percentile=0.5):
    #index_median = sum(( df_zip["N1"]/ sum(df_zip["N1"])).cumsum() <= percentile)
    #val_below_or_at_median = (df_zip["A00100"] /df_zip["N1"]).iloc[index_median]
    #return val_below_or_at_median

In [18]:
tax_data_valid_zip_cleaned = tax_data_valid_zip_cleaned.reset_index(drop=True)
tax_data_valid_zip_cleaned

STATEFIPS STATE  ZIPCODE  AGI_STUB      N1   MARS1   MARS2  MARS4  \
0             1.0    AL  35004.0       1.0  1490.0   970.0   230.0  280.0   
1             1.0    AL  35004.0       2.0  1350.0   630.0   360.0  300.0   
2             1.0    AL  35004.0       3.0   970.0   310.0   490.0  140.0   
3             1.0    AL  35004.0       4.0   620.0   110.0   470.0   30.0   
4             1.0    AL  35004.0       5.0   620.0    40.0   560.0   20.0   
...           ...   ...      ...       ...     ...     ...     ...    ...   
166248       56.0    WY  99999.0       2.0  3680.0  2000.0  1080.0  490.0   
166249       56.0    WY  99999.0       3.0  2630.0  1100.0  1240.0  250.0   
166250       56.0    WY  99999.0       4.0  2020.0   590.0  1310.0  100.0   
166251       56.0    WY  99999.0       5.0  2670.0   430.0  2170.0   90.0   
166252       56.0    WY  99999.0       6.0   760.0   140.0   650.0    0.0   

          PREP      N2  ...  N10300    A10300  N85530  A85530  N85300  A85300  \
0        700.0  2160.0  ...   690.0     610.0     0.0     0.0     0.0     0.0   
1        610.0  2540.0  ...  1140.0    3019.0     0.0     0.0     0.0     0.0   
2        450.0  2160.0  ...   930.0    5009.0     0.0     0.0     0.0     0.0   
3        300.0  1610.0  ...   620.0    5190.0     0.0     0.0     0.0     0.0   
4        320.0  1770.0  ...   620.0   10129.0     0.0     0.0     0.0     0.0   
...        ...     ...  ...     ...       ...     ...     ...     ...     ...   
166248  2020.0  6660.0  ...  3220.0    9235.0     0.0     0.0     0.0     0.0   
166249  1610.0  5440.0  ...  2530.0   15579.0     0.0     0.0     0.0     0.0   
166250  1330.0  4780.0  ...  1990.0   19278.0     0.0     0.0     0.0     0.0   
166251  1850.0  6930.0  ...  2650.0   49853.0     0.0     0.0     0.0     0.0   
166252   720.0  1890.0  ...   760.0  139792.0   170.0   198.0   490.0  3842.0   

        N11901   A11901  N11902   A11902  
0        120.0     94.0  1290.0   2792.0  
1        210.0    301.0  1130.0   2935.0  
2        200.0    376.0   760.0   2058.0  
3        150.0    374.0   460.0   1390.0  
4        220.0    854.0   410.0   1709.0  
...        ...      ...     ...      ...  
166248   550.0    869.0  2960.0   7717.0  
166249   560.0   1273.0  1890.0   5330.0  
166250   470.0   1635.0  1390.0   4696.0  
166251   950.0   5576.0  1460.0   6853.0  
166252   340.0  14487.0   160.0  21167.0  

[166253 rows x 91 columns]

In [19]:
def compute_percentile_zip(df_zip, percentile=0.65):
    index_median = sum(( df_zip["N1"]/ sum(df_zip["N1"])).cumsum() <= percentile)
    val_below_or_at_median = (df_zip["A00100"] /df_zip["N1"]).iloc[index_median]
    return val_below_or_at_median

In [20]:
## WRITE YOUR CODE HERE
grouper = tax_data_valid_zip_cleaned.groupby('ZIPCODE').apply(compute_percentile_zip)

In [21]:
zip_rev_all = grouper.sort_values(ascending=False)
zip_rev_all.head()

ZIPCODE
33109.0    3954.114286
33480.0    3413.301538
94301.0    3109.443711
94027.0    3091.537013
10577.0    2414.855556
dtype: float64

- What are the three zip codes with the most significant 65th percentile value for income?

In [22]:
## WRITE YOUR CODE HERE
zip_rev_all.head(3)

ZIPCODE
33109.0    3954.114286
33480.0    3413.301538
94301.0    3109.443711
dtype: float64

# 2 Working with the Medicaid Data

### Loading and exploring the data 

* Load the Medicaid data stored in the file `medicaid_data.csv` into a `DataFrame` called `medicaid_data`. The file is located in the `data` directory of the assignment folder. 
* Note that this is quite large and may take some time to load on a computer with modest RAM resources (4GB or less)


In [23]:
## WRITE YOUR CODE HERE
medicaid_data = pd.read_csv('data/medicaid_data.csv')
medicaid_data.head()

Utilization Type State          NDC Product Name  Units Reimbursed  \
0             MCOU    PA  55150023930   Dexamethas              33.0   
1             FFSU    NY     23917710   ALPHAGAN P             570.0   
2             MCOU    OR  13925050501   Dapsone 10             456.0   
3             FFSU    MN  51862006401     DIAZEPAM             780.0   
4             FFSU    MN    781237101   DEXTROAMPH             451.0   

   Number of Prescriptions  Total Amount Reimbursed  \
0                     19.0                   234.98   
1                     57.0                 16006.34   
2                     15.0                  1052.42   
3                     16.0                    89.60   
4                     12.0                  1411.24   

   Medicaid Amount Reimbursed  Non Medicaid Amount Reimbursed  \
0                      234.98                            0.00   
1                    16006.34                            0.00   
2                     1052.42                            0.00   
3                       77.60                           12.00   
4                      198.93                         1212.31   

               Location  
0    (40.5773, -77.264)  
1   (42.1497, -74.9384)  
2  (44.5672, -122.1269)  
3   (45.7326, -93.9196)  
4   (45.7326, -93.9196)


- Modify `medicaid_data` to uppercase all the column names 

  - If your solution uses an assignment, the righthand side of the assignment (rvalue) can only use `DataFrame` or `Series` methods or properties


In [24]:
## WRITE YOUR CODE HERE
medicaid_data.columns = medicaid_data.columns.str.upper()
medicaid_data.head()

UTILIZATION TYPE STATE          NDC PRODUCT NAME  UNITS REIMBURSED  \
0             MCOU    PA  55150023930   Dexamethas              33.0   
1             FFSU    NY     23917710   ALPHAGAN P             570.0   
2             MCOU    OR  13925050501   Dapsone 10             456.0   
3             FFSU    MN  51862006401     DIAZEPAM             780.0   
4             FFSU    MN    781237101   DEXTROAMPH             451.0   

   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
0                     19.0                   234.98   
1                     57.0                 16006.34   
2                     15.0                  1052.42   
3                     16.0                    89.60   
4                     12.0                  1411.24   

   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
0                      234.98                            0.00   
1                    16006.34                            0.00   
2                     1052.42                            0.00   
3                       77.60                           12.00   
4                      198.93                         1212.31   

               LOCATION  
0    (40.5773, -77.264)  
1   (42.1497, -74.9384)  
2  (44.5672, -122.1269)  
3   (45.7326, -93.9196)  
4   (45.7326, -93.9196)


- Familiarize your self with the data
  - the `NDC` column stands for National Drug Code, a universal product identifier for human drugs in the United States
  
  - The remaining column names are self-explanatory
  
- Explore the number of lines and columns in the data

- Check that your column headers are in uppercase

In [25]:
## WRITE YOUR CODE HERE
print(medicaid_data.columns)
print(medicaid_data.shape)
print(medicaid_data.size)

Index(['UTILIZATION TYPE', 'STATE', 'NDC', 'PRODUCT NAME', 'UNITS REIMBURSED',
       'NUMBER OF PRESCRIPTIONS', 'TOTAL AMOUNT REIMBURSED',
       'MEDICAID AMOUNT REIMBURSED', 'NON MEDICAID AMOUNT REIMBURSED',
       'LOCATION'],
      dtype='object')
(1695546, 10)
16955460


* If you explore the  `Location` column for all the entries for which "STATE" value is equal to "HI" you'll notice that all the values are identical

* Are there any states that have more than one value for `Location`. 
  * Hint: think about using a sorted aggregation as part of a split-apply-combine operation to answer this question
  - Your answer can only use `DataFrame` or `Series` methods or properties


In [26]:
hawaii = medicaid_data.groupby('STATE').get_group('HI')
hawaii

UTILIZATION TYPE STATE          NDC PRODUCT NAME  UNITS REIMBURSED  \
154                 MCOU    HI  16714034804   DASETTA 1-             448.0   
294                 MCOU    HI  49348002972    SM TRIPLE             821.6   
318                 MCOU    HI    378043301    BUPROPION            5461.0   
397                 MCOU    HI  49348004537   SM ALLERGY            5016.0   
557                 MCOU    HI     54007928   BALSALAZID            3630.0   
...                  ...   ...          ...          ...               ...   
1694941             MCOU    HI  45802006436   TRIAMCINOL           11520.0   
1695124             MCOU    HI    299590825   EPIDUO 0.1            1890.0   
1695291             MCOU    HI  69097084507   CYCLOBENZA            4811.0   
1695327             MCOU    HI  52544024928   MICROGESTI            9576.0   
1695509             MCOU    HI     93227534    AMOX-CLAV            3493.0   

         NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
154                         16.0                   267.07   
294                         27.0                   143.34   
318                         97.0                  2206.56   
397                         35.0                   105.57   
557                         14.0                  3204.75   
...                          ...                      ...   
1694941                    138.0                  1331.15   
1695124                     37.0                  6509.69   
1695291                    129.0                   714.83   
1695327                    131.0                  5166.67   
1695509                    185.0                  2287.13   

         MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
154                          267.07                            0.00   
294                          143.34                            0.00   
318                         2206.56                            0.00   
397                          105.57                            0.00   
557                         3204.75                            0.00   
...                             ...                             ...   
1694941                     1331.15                            0.00   
1695124                     6509.69                            0.00   
1695291                      707.49                            7.34   
1695327                     5166.67                            0.00   
1695509                     2282.69                            4.44   

                     LOCATION  
154      (21.1098, -157.5311)  
294      (21.1098, -157.5311)  
318      (21.1098, -157.5311)  
397      (21.1098, -157.5311)  
557      (21.1098, -157.5311)  
...                       ...  
1694941  (21.1098, -157.5311)  
1695124  (21.1098, -157.5311)  
1695291  (21.1098, -157.5311)  
1695327  (21.1098, -157.5311)  
1695509  (21.1098, -157.5311)  

[13438 rows x 10 columns]

In [27]:
def loc_comp(df):
    df['match'] = df.LOCATION.eq(df.LOCATION.shift())
    return df

In [28]:
groups = medicaid_data.groupby('STATE').apply(loc_comp)

In [29]:
new_df = groups[groups['match'] == False]
counts = new_df.value_counts(new_df['STATE'])
counts

STATE
XX    153253
AL         1
ND         1
NE         1
NH         1
NJ         1
NM         1
NV         1
NY         1
OH         1
OK         1
OR         1
PA         1
RI         1
SC         1
SD         1
TN         1
TX         1
UT         1
VA         1
VT         1
WA         1
WI         1
WV         1
WY         1
NC         1
AK         1
MS         1
IA         1
AR         1
AZ         1
CA         1
CO         1
CT         1
DC         1
DE         1
FL         1
GA         1
HI         1
ID         1
MO         1
IL         1
IN         1
KS         1
KY         1
LA         1
MA         1
MD         1
ME         1
MI         1
MN         1
MT         1
dtype: int64



* To compare medication prescriptions across states in a fair and balanced way, we need the number of Medicaid beneficiaries in each state. The following example illustrates the importance of normalizing the values `UNITS REIMBURSED` for each medication in each state by the number of Medicaid enrollees in each state.
  
* The `medicaid_data` DataFrame shows that for the drug with NDC `61958180101` (the drug name is HARVONI and it's used to treat Hepatitis C) there were 11,886  units sold in KY, versus 40,142 in CA -- that's almost 4 times more units sold in CA compared to KY. However, there are 1,284,193 Medicaid enrollees in KY, versus 13,096,861 in California. If we normalize the number of units sold in KY, versus CA, we find that the normalized there were close to 3 times more HARVONI prescription in KY  than in CA. This is ___perhaps___ justified by the fact the KY has one of the highest rates of reported cases of Hepatitis C in the US (2.7% in KY versus 0.2% in CA).
  
https://www.cdc.gov/hepatitis/statistics/2015surveillance/pdfs/2015hepsurveillancerpt.pdf

* The number of enrollees per state was obtained here:

    https://www.medicaid.gov/medicaid/managed-care/enrollment/index.html
    
    
* A parsed/processed version (medicaid_enrollment.tsv) can be in data director of the assignment folder. Use `pandas` to load the medicaid_enrollmen file into DataFrame named `medicaid_enrollment`

In [30]:
## WRITE YOUR CODE HERE
medicaid_enrollment = pd.read_csv('data/medicaid_enrollment.tsv', sep='\t')
medicaid_enrollment

STATE Total Medicaid Enrollees
0                    Alabama                1,050,989
1                     Alaska                  164,783
2             American Samoa                      NaN
3                    Arizona                1,740,520
4                   Arkansas                  762,166
5                 California               13,096,861
6                   Colorado                1,264,600
7                Connecticut                  746,119
8                   Delaware                  227,909
9       District of Columbia                  271,428
10                   Florida                3,808,334
11                   Georgia                1,990,810
12                      Guam                      n/a
13                    Hawaii                  340,513
14                     Idaho                  283,355
15                  Illinois                3,269,999
16                   Indiana                1,295,358
17                      Iowa                  618,505
18                    Kansas                  403,844
19                  Kentucky                1,284,193
20                 Louisiana                1,402,212
21                     Maine                  288,324
22                  Maryland                1,271,445
23             Massachusetts                1,829,618
24                  Michigan                3,947,031
25                 Minnesota                1,052,521
26               Mississippi                  740,937
27                  Missouri                  944,257
28                   Montana                  139,950
29                  Nebraska                  239,463
30                    Nevada                  588,304
31             New Hampshire                  186,399
32                New Jersey                1,705,594
33                New Mexico                  826,155
34                  New York                6,281,038
35            North Carolina                1,965,805
36              North Dakota                   86,250
37  Northern Mariana Islands                      NaN
38                      Ohio                3,060,446
39                  Oklahoma                  829,561
40                    Oregon                1,123,913
41              Pennsylvania                2,569,232
42               Puerto Rico                1,458,819
43              Rhode Island                  308,521
44            South Carolina                1,233,430
45              South Dakota                  124,497
46                 Tennessee                1,562,745
47                     Texas                4,273,982
48                      Utah                  293,867
49                   Vermont                  206,469
50            Virgin Islands                      NaN
51                  Virginia                1,092,225
52                Washington                1,771,679
53             West Virginia                  545,748
54                 Wisconsin                1,209,714
55                   Wyoming                   66,532

* Modify `medicaid_enrollment` to uppercase all the column names 

  * Your answer can only use `DataFrame` or `Series` methods or properties
  * Do not hardcode the operation by uppercasing the columns yourself


In [31]:
## WRITE YOUR CODE HERE
medicaid_enrollment.columns = medicaid_enrollment.columns.str.upper()
medicaid_enrollment.columns

Index(['STATE', 'TOTAL MEDICAID ENROLLEES'], dtype='object')

* Note that some states/territories have missing values. Remove the missing values and save the resulting `DataFrame` as a new variable named `medicaid_enrollment_cleaned`

* Pay attention to how 'n/a' is given here!
* After cleaning, do you still have the Guam entry? If so, reconsider what missing values means in this context

In [32]:
## WRITE YOUR CODE HERE
medicaid_enrollment_cleaned = medicaid_enrollment.dropna()
medicaid_enrollment_cleaned = medicaid_enrollment_cleaned[medicaid_enrollment_cleaned['TOTAL MEDICAID ENROLLEES'].str.contains('n/a') == False]
medicaid_enrollment_cleaned

STATE TOTAL MEDICAID ENROLLEES
0                Alabama                1,050,989
1                 Alaska                  164,783
3                Arizona                1,740,520
4               Arkansas                  762,166
5             California               13,096,861
6               Colorado                1,264,600
7            Connecticut                  746,119
8               Delaware                  227,909
9   District of Columbia                  271,428
10               Florida                3,808,334
11               Georgia                1,990,810
13                Hawaii                  340,513
14                 Idaho                  283,355
15              Illinois                3,269,999
16               Indiana                1,295,358
17                  Iowa                  618,505
18                Kansas                  403,844
19              Kentucky                1,284,193
20             Louisiana                1,402,212
21                 Maine                  288,324
22              Maryland                1,271,445
23         Massachusetts                1,829,618
24              Michigan                3,947,031
25             Minnesota                1,052,521
26           Mississippi                  740,937
27              Missouri                  944,257
28               Montana                  139,950
29              Nebraska                  239,463
30                Nevada                  588,304
31         New Hampshire                  186,399
32            New Jersey                1,705,594
33            New Mexico                  826,155
34              New York                6,281,038
35        North Carolina                1,965,805
36          North Dakota                   86,250
38                  Ohio                3,060,446
39              Oklahoma                  829,561
40                Oregon                1,123,913
41          Pennsylvania                2,569,232
42           Puerto Rico                1,458,819
43          Rhode Island                  308,521
44        South Carolina                1,233,430
45          South Dakota                  124,497
46             Tennessee                1,562,745
47                 Texas                4,273,982
48                  Utah                  293,867
49               Vermont                  206,469
51              Virginia                1,092,225
52            Washington                1,771,679
53         West Virginia                  545,748
54             Wisconsin                1,209,714
55               Wyoming                   66,532

### Converting `TOTAL MEDICAID ENROLLEE` Data Type

* Given that data on `TOTAL MEDICAID ENROLLEE` column contains commas on file (ex. 3,269,999 instead of 3269999), `pandas` has erroneously set the data type for that column as a string. We need to convert the column from string to `int` since we will be using it in an arithmetic expression during normalization

 

* Inspect the `dtype` property of "TOTAL MEDICAID ENROLLEES" column, and  make sure that the data type is `int`


In [33]:
## WRITE YOUR CODE HERE
medicaid_enrollment_cleaned['TOTAL MEDICAID ENROLLEES'] = medicaid_enrollment_cleaned['TOTAL MEDICAID ENROLLEES'].str.replace(',', '').astype(float)
medicaid_enrollment_cleaned

STATE  TOTAL MEDICAID ENROLLEES
0                Alabama                 1050989.0
1                 Alaska                  164783.0
3                Arizona                 1740520.0
4               Arkansas                  762166.0
5             California                13096861.0
6               Colorado                 1264600.0
7            Connecticut                  746119.0
8               Delaware                  227909.0
9   District of Columbia                  271428.0
10               Florida                 3808334.0
11               Georgia                 1990810.0
13                Hawaii                  340513.0
14                 Idaho                  283355.0
15              Illinois                 3269999.0
16               Indiana                 1295358.0
17                  Iowa                  618505.0
18                Kansas                  403844.0
19              Kentucky                 1284193.0
20             Louisiana                 1402212.0
21                 Maine                  288324.0
22              Maryland                 1271445.0
23         Massachusetts                 1829618.0
24              Michigan                 3947031.0
25             Minnesota                 1052521.0
26           Mississippi                  740937.0
27              Missouri                  944257.0
28               Montana                  139950.0
29              Nebraska                  239463.0
30                Nevada                  588304.0
31         New Hampshire                  186399.0
32            New Jersey                 1705594.0
33            New Mexico                  826155.0
34              New York                 6281038.0
35        North Carolina                 1965805.0
36          North Dakota                   86250.0
38                  Ohio                 3060446.0
39              Oklahoma                  829561.0
40                Oregon                 1123913.0
41          Pennsylvania                 2569232.0
42           Puerto Rico                 1458819.0
43          Rhode Island                  308521.0
44        South Carolina                 1233430.0
45          South Dakota                  124497.0
46             Tennessee                 1562745.0
47                 Texas                 4273982.0
48                  Utah                  293867.0
49               Vermont                  206469.0
51              Virginia                 1092225.0
52            Washington                 1771679.0
53         West Virginia                  545748.0
54             Wisconsin                 1209714.0
55               Wyoming                   66532.0

### Associating `medicaid_data` and `medicaid_enrollment_cleaned`

- We can use the shared State information across both tables to associate both tables (SQL JOIN).
- However,  `medicaid_data` contains two-letter state abbreviations, while `medicaid_enrollment_cleaned` contains the complete state name
  - We need to convert (or append) two-letter state abbreviations to `medicaid_enrollment_cleaned`

- Pandas can read HTML and parse the code for tables. We will use that functionality to read in the state abbreviations from a Wikipedia page.
  - A brief description of what the code does is included in the comments

In [34]:
import requests

url = 'https://www.50states.com/abbreviations.htm'
header = {
    "User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.75 Safari/537.36",
    "X-Requested-With": "XMLHttpRequest"
}

r = requests.get(url, headers=header)

tables = pd.read_html(r.text)


# We access the desired table by giving it's index.
# Since the URL contain only one table, then we can access that table using index 0
Codes_abbreviations = tables[0]
Codes_abbreviations.head(5)

US STATE POSTAL ABBREVIATION STANDARD ABBREVIATION
0     Alabama                  AL                  Ala.
1      Alaska                  AK                Alaska
2     Arizona                  AZ                 Ariz.
3    Arkansas                  AR                  Ark.
4  California                  CA                Calif.

* Change the the `DataFrame`'s headers from ['US State:', 'Abbreviation:'] to ['US STATE', 'ABBREVIATION']

  * You can hard code this operation


In [35]:
## WRITE YOUR CODE HERE 
Codes_abbreviations = Codes_abbreviations.rename(columns={'POSTAL ABBREVIATION': 'ABBREVIATION'})
Codes_abbreviations

US STATE ABBREVIATION STANDARD ABBREVIATION
0          Alabama           AL                  Ala.
1           Alaska           AK                Alaska
2          Arizona           AZ                 Ariz.
3         Arkansas           AR                  Ark.
4       California           CA                Calif.
5         Colorado           CO                Color.
6      Connecticut           CT                 Conn.
7         Delaware           DE                  Del.
8          Florida           FL                  Fla.
9          Georgia           GA                   Ga.
10          Hawaii           HI                Hawaii
11           Idaho           ID                 Idaho
12        Illinois           IL                  Ill.
13         Indiana           IN                  Ind.
14            Iowa           IA                  Iowa
15          Kansas           KS                  Kan.
16        Kentucky           KY                   Ky.
17       Louisiana           LA                   La.
18           Maine           ME                 Maine
19        Maryland           MD                   Md.
20   Massachusetts           MA                 Mass.
21        Michigan           MI                 Mich.
22       Minnesota           MN                 Minn.
23     Mississippi           MS                 Miss.
24        Missouri           MO                   Mo.
25         Montana           MT                 Mont.
26        Nebraska           NE                  Neb.
27          Nevada           NV                  Nev.
28   New Hampshire           NH                  N.H.
29      New Jersey           NJ                  N.J.
30      New Mexico           NM                  N.M.
31        New York           NY                  N.Y.
32  North Carolina           NC                  N.C.
33    North Dakota           ND                  N.D.
34            Ohio           OH                  Ohio
35        Oklahoma           OK                 Okla.
36          Oregon           OR                  Ore.
37    Pennsylvania           PA                   Pa.
38    Rhode Island           RI                  R.I.
39  South Carolina           SC                  S.C.
40    South Dakota           SD                S.Dak.
41       Tennessee           TN                 Tenn.
42           Texas           TX                  Tex.
43            Utah           UT                  Utah
44         Vermont           VT                  V.T.
45        Virginia           VA                   Va.
46      Washington           WA                 Wash.
47   West Virginia           WV                 W.Va.
48       Wisconsin           WI                  Wis.
49         Wyoming           WY                  Wyo.

* Combine the tables `medicaid_enrollment_cleaned` and `Codes_abbreviations` such that the resulting `DataFrame` contains all the columns in `medicaid_enrollment_cleaned` and only `ABBREVIATION` from `Codes_abbreviations` 
* Save the results to variable named `medicaid_enrollment_cleaned_with_zip`

- `medicaid_enrollment_cleaned_with_zip` should look like the following ( '...' represents remaining data that is not shown):


```
   STATE    Total Medicaid Enrollees    ABBREVIATION
0    Alabama    1,050,989    AL
1    Alaska    164,783    AK
2    Arizona    1,740,520    AZ
3    Arkansas    762,166    AR
4    California    13,096,861    CA
...
```

* We did not cover joins in class -- you find a plethora of examples on how to do this online. See for instance:

`https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.merge.html`

* If you cannot get it to work, contact the `TA` for the solution. You will not be penalized if you don't answer this question. 

In [36]:
medicaid_enrollment_cleaned_with_zip = medicaid_enrollment_cleaned.merge(Codes_abbreviations, how='left', left_on='STATE', right_on='US STATE')

* We have no further use for the column STATE in  `medicaid_enrollment_cleaned_with_zip`
  * Remove the column make sure your data in `medicaid_enrollment_cleaned_with_zip` looks like the following  ( `...` represents remaining data that is not shown):

```
    Total Medicaid Enrollees    ABBREVIATION
0   1,050,989                  AL
1   164,783                    AK
2   1,740,520                  AZ
3   762,166                    AR
4   13,096,861                 CA
....
```

In [37]:
medicaid_enrollment_cleaned_with_zip = medicaid_enrollment_cleaned_with_zip.drop(['US STATE',
                                                                                  'STANDARD ABBREVIATION',
                                                                                  'STATE'], axis=1)
medicaid_enrollment_cleaned_with_zip

TOTAL MEDICAID ENROLLEES ABBREVIATION
0                  1050989.0           AL
1                   164783.0           AK
2                  1740520.0           AZ
3                   762166.0           AR
4                 13096861.0           CA
5                  1264600.0           CO
6                   746119.0           CT
7                   227909.0           DE
8                   271428.0          NaN
9                  3808334.0           FL
10                 1990810.0           GA
11                  340513.0           HI
12                  283355.0           ID
13                 3269999.0           IL
14                 1295358.0           IN
15                  618505.0           IA
16                  403844.0           KS
17                 1284193.0           KY
18                 1402212.0           LA
19                  288324.0           ME
20                 1271445.0           MD
21                 1829618.0           MA
22                 3947031.0           MI
23                 1052521.0           MN
24                  740937.0           MS
25                  944257.0           MO
26                  139950.0           MT
27                  239463.0           NE
28                  588304.0           NV
29                  186399.0           NH
30                 1705594.0           NJ
31                  826155.0           NM
32                 6281038.0           NY
33                 1965805.0           NC
34                   86250.0           ND
35                 3060446.0           OH
36                  829561.0           OK
37                 1123913.0           OR
38                 2569232.0           PA
39                 1458819.0          NaN
40                  308521.0           RI
41                 1233430.0           SC
42                  124497.0           SD
43                 1562745.0           TN
44                 4273982.0           TX
45                  293867.0           UT
46                  206469.0           VT
47                 1092225.0           VA
48                 1771679.0           WA
49                  545748.0           WV
50                 1209714.0           WI
51                   66532.0           WY

- Use `DataFrame medicaid_enrollment_cleaned_with_zip` to assign the appropriate number of Medicaid enrollees to each entry in the `medicaid_data`

   I.E., instead of the 10 original columns, `medicaid_data` will now have an 11th column representing the `TOTAL MEDICAID ENROLLEES` according to the STATE value in the entry.
   
- Save the resulting DataFrame into a new variable called `medicaid_data_w_enrollments`
- The resulting DataFrame should look like the following (`...` represents remaining data that is not shown):

```
UTILIZATION TYPE    STATE    NDC    PRODUCT NAME    UNITS REIMBURSED    NUMBER OF PRESCRIPTIONS    TOTAL AMOUNT REIMBURSED    MEDICAID AMOUNT REIMBURSED    NON MEDICAID AMOUNT REIMBURSED    LOCATION
0    MCOU    PA    55150023930    Dexamethas    33.0    19.0    234.98    234.98    0.0    (40.5773, -77.264)
1    FFSU    NY    23917710    ALPHAGAN P    570.0    57.0    16006.34    16006.34    0.0    (42.1497, -74.9384)
2    MCOU    OR    13925050501    Dapsone 10    456.0    15.0    1052.42    1052.42    0.0    (44.5672, -122.1269)
...
```

* The order of the columns in the `DataFrame` is not important. This answer uses the same approach as the one used to `merge` the tables above. 

In [38]:
## WRITE YOUR CODE HERE
#medicaid_enrollment_cleaned_with_zip = medicaid_enrollment_cleaned.merge(Codes_abbreviations, how='left', #left_on='STATE', right_on='US STATE')
medicaid_data_w_enrollments = medicaid_data.merge(medicaid_enrollment_cleaned_with_zip,
                                                  how='left', left_on='STATE', right_on='ABBREVIATION')

In [39]:
medicaid_data_w_enrollments = medicaid_data_w_enrollments.drop(['ABBREVIATION'], axis=1)

- Remove any lines where "STATE" or "PRODUCT NAME" are missing from  `medicaid_data_w_enrollments`

In [40]:
## WRITE YOUR CODE HERE
medicaid_data_w_enrollments[medicaid_data_w_enrollments['STATE'].str.strip().astype(bool)]
medicaid_data_w_enrollments[medicaid_data_w_enrollments['PRODUCT NAME'].str.strip().astype(bool)]
medicaid_data_w_enrollments.dropna(subset=['STATE', 'PRODUCT NAME'], inplace=True)
medicaid_data_w_enrollments = medicaid_data_w_enrollments[~medicaid_data_w_enrollments.STATE.str.contains('XX')]
medicaid_data_w_enrollments

UTILIZATION TYPE STATE          NDC PRODUCT NAME  UNITS REIMBURSED  \
0                   MCOU    PA  55150023930   Dexamethas              33.0   
1                   FFSU    NY     23917710   ALPHAGAN P             570.0   
2                   MCOU    OR  13925050501   Dapsone 10             456.0   
3                   FFSU    MN  51862006401     DIAZEPAM             780.0   
4                   FFSU    MN    781237101   DEXTROAMPH             451.0   
...                  ...   ...          ...          ...               ...   
1695541             FFSU    CA  65862056230   OLANZAPINE            1018.0   
1695542             MCOU    PA    378700393   PAROXETINE            8227.0   
1695543             MCOU    PA  66685101201    AMOX CLAV          167146.0   
1695544             MCOU    NJ  50242004164   CATHFLO AC             182.5   
1695545             MCOU    NM    378021301   CHLORTHALI            2465.0   

         NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
0                           19.0                   234.98   
1                           57.0                 16006.34   
2                           15.0                  1052.42   
3                           16.0                    89.60   
4                           12.0                  1411.24   
...                          ...                      ...   
1695541                     19.0                   357.29   
1695542                    215.0                  1815.23   
1695543                   1283.0                 26356.28   
1695544                    155.0                 16620.63   
1695545                     66.0                  2586.96   

         MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
0                            234.98                            0.00   
1                          16006.34                            0.00   
2                           1052.42                            0.00   
3                             77.60                           12.00   
4                            198.93                         1212.31   
...                             ...                             ...   
1695541                      357.29                            0.00   
1695542                     1813.44                            1.79   
1695543                    25638.49                          717.79   
1695544                    15103.73                         1516.90   
1695545                     2586.96                            0.00   

                     LOCATION  TOTAL MEDICAID ENROLLEES  
0          (40.5773, -77.264)                 2569232.0  
1         (42.1497, -74.9384)                 6281038.0  
2        (44.5672, -122.1269)                 1123913.0  
3         (45.7326, -93.9196)                 1052521.0  
4         (45.7326, -93.9196)                 1052521.0  
...                       ...                       ...  
1695541    (36.17, -119.7462)                13096861.0  
1695542    (40.5773, -77.264)                 2569232.0  
1695543    (40.5773, -77.264)                 2569232.0  
1695544    (40.314, -74.5089)                 1705594.0  
1695545  (34.8375, -106.2371)                  826155.0  

[1542292 rows x 11 columns]

- Use ["STATE", "PRODUCT NAME"] as hierarchical index for `medicaid_data_w_enrollments`. Recall that a hierarchical index is simply an index with multiple levels of indexing (multiple columns)
  * Hint: the function to set an index on a `DataFrame` can take a single column name or a list of column names. The list here is  ["STATE", "NDC"]
- Call the new data `medicaid_data_w_enrollments_hierarch`
- Inspect your data to make sure the new index has now two levels (STATE and NDC)

In [41]:
## WRITE YOUR CODE HERE 
medicaid_data_w_enrollments_hierarch = medicaid_data_w_enrollments.set_index(['STATE', 'NDC'])

In [42]:
medicaid_data_w_enrollments_hierarch

UTILIZATION TYPE PRODUCT NAME  UNITS REIMBURSED  \
STATE NDC                                                           
PA    55150023930             MCOU   Dexamethas              33.0   
NY    23917710                FFSU   ALPHAGAN P             570.0   
OR    13925050501             MCOU   Dapsone 10             456.0   
MN    51862006401             FFSU     DIAZEPAM             780.0   
      781237101               FFSU   DEXTROAMPH             451.0   
...                            ...          ...               ...   
CA    65862056230             FFSU   OLANZAPINE            1018.0   
PA    378700393               MCOU   PAROXETINE            8227.0   
      66685101201             MCOU    AMOX CLAV          167146.0   
NJ    50242004164             MCOU   CATHFLO AC             182.5   
NM    378021301               MCOU   CHLORTHALI            2465.0   

                   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
STATE NDC                                                             
PA    55150023930                     19.0                   234.98   
NY    23917710                        57.0                 16006.34   
OR    13925050501                     15.0                  1052.42   
MN    51862006401                     16.0                    89.60   
      781237101                       12.0                  1411.24   
...                                    ...                      ...   
CA    65862056230                     19.0                   357.29   
PA    378700393                      215.0                  1815.23   
      66685101201                   1283.0                 26356.28   
NJ    50242004164                    155.0                 16620.63   
NM    378021301                       66.0                  2586.96   

                   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
STATE NDC                                                                       
PA    55150023930                      234.98                            0.00   
NY    23917710                       16006.34                            0.00   
OR    13925050501                     1052.42                            0.00   
MN    51862006401                       77.60                           12.00   
      781237101                        198.93                         1212.31   
...                                       ...                             ...   
CA    65862056230                      357.29                            0.00   
PA    378700393                       1813.44                            1.79   
      66685101201                    25638.49                          717.79   
NJ    50242004164                    15103.73                         1516.90   
NM    378021301                       2586.96                            0.00   

                               LOCATION  TOTAL MEDICAID ENROLLEES  
STATE NDC                                                          
PA    55150023930    (40.5773, -77.264)                 2569232.0  
NY    23917710      (42.1497, -74.9384)                 6281038.0  
OR    13925050501  (44.5672, -122.1269)                 1123913.0  
MN    51862006401   (45.7326, -93.9196)                 1052521.0  
      781237101     (45.7326, -93.9196)                 1052521.0  
...                                 ...                       ...  
CA    65862056230    (36.17, -119.7462)                13096861.0  
PA    378700393      (40.5773, -77.264)                 2569232.0  
      66685101201    (40.5773, -77.264)                 2569232.0  
NJ    50242004164    (40.314, -74.5089)                 1705594.0  
NM    378021301    (34.8375, -106.2371)                  826155.0  

[1542292 rows x 9 columns]


* Write a single Pandas expression to print all the lines with containing NDC 61958180101 in "PA"

 * Use a single indexing call (bracket notation) using `loc`

 * Hint 1: Since your index is hierarchical, `loc` is expecting two values, the first for STATE and the second for NDC


In [43]:
medicaid_data_w_enrollments_hierarch.loc['PA', 61958180101]

/var/folders/jn/lb25bnzx4js5f2jfpwzh8brw0000gn/T/ipykernel_74618/4092694910.py:1: PerformanceWarning: indexing past lexsort depth may impact performance.
  medicaid_data_w_enrollments_hierarch.loc['PA', 61958180101]


UTILIZATION TYPE PRODUCT NAME  UNITS REIMBURSED  \
STATE NDC                                                           
PA    61958180101             FFSU   Harvoni  (             924.0   
      61958180101             MCOU   Harvoni  (            2604.0   
      61958180101             FFSU   Harvoni  (            1008.0   
      61958180101             MCOU   Harvoni  (            1932.0   
      61958180101             FFSU   Harvoni  (             924.0   
      61958180101             MCOU   Harvoni  (            3220.0   

                   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
STATE NDC                                                             
PA    61958180101                     33.0               1017644.54   
      61958180101                     93.0               2952891.16   
      61958180101                     36.0               1107725.22   
      61958180101                     69.0               2178602.30   
      61958180101                     33.0               1015383.60   
      61958180101                    115.0               3599192.35   

                   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
STATE NDC                                                                       
PA    61958180101                   985269.84                        32374.70   
      61958180101                  2860100.54                        92790.62   
      61958180101                  1107725.22                            0.00   
      61958180101                  2118612.30                        59990.00   
      61958180101                  1015383.60                            0.00   
      61958180101                  3482188.68                       117003.67   

                             LOCATION  TOTAL MEDICAID ENROLLEES  
STATE NDC                                                        
PA    61958180101  (40.5773, -77.264)                 2569232.0  
      61958180101  (40.5773, -77.264)                 2569232.0  
      61958180101  (40.5773, -77.264)                 2569232.0  
      61958180101  (40.5773, -77.264)                 2569232.0  
      61958180101  (40.5773, -77.264)                 2569232.0  
      61958180101  (40.5773, -77.264)                 2569232.0

In [44]:
ratio_reimbursed = (medicaid_data_w_enrollments_hierarch.groupby(['STATE', 'NDC'])['UNITS REIMBURSED'].sum() /                             medicaid_data_w_enrollments_hierarch.groupby(['STATE', 'NDC'])['TOTAL MEDICAID ENROLLEES'].unique())

In [45]:
ratio_reimbursed = ratio_reimbursed.astype(float)
medicaid_reimbursement_per_enrollee = np.log2(ratio_reimbursed)
medicaid_reimbursement_per_enrollee

STATE  NDC        
AK     2143380        -9.609109
       2143480       -10.008280
       2322730        -6.109830
       2322830        -4.444321
       2322930        -3.855995
                        ...    
WY     76329336901    -9.773833
       76385010301    -6.146779
       76385010401    -4.707744
       76385010501    -5.965123
       76439030910    -4.794145
Length: 473016, dtype: float64

- To facilitate working with the final data, we are going to unstack `medicaid_reimbursement_per_enrollee` into a variable called  `medicaid_norm_ndc`

- Using `medicaid_reimbursement_per_enrollee`, generate a `DataFrame` where: 
  - index should be the two-letter state symbol 
  - the column names should be the NDC codes 

- The `DataFrame`  should be formatted as in the image below
  - Hint, simply unstack the data


<img src="media/unstacked.png" alt="drawing" style="width:900px;"/>


In [46]:
## WRITE YOUR CODE HERE 
medicaid_norm_ndc = medicaid_reimbursement_per_enrollee.unstack(level=-1)

In [47]:
medicaid_norm_ndc

NDC    2143380      2143480      2144511      2144527      2197590      \
STATE                                                                    
AK       -9.609109   -10.008280          NaN          NaN          NaN   
AL       -9.940595    -9.805485          NaN          NaN          NaN   
AR      -12.985157   -12.657103          NaN          NaN          NaN   
AZ      -11.533870   -10.821194          NaN          NaN          NaN   
CA      -10.926863   -10.698372   -16.127018          NaN    -9.464364   
CO      -10.862982    -9.925954          NaN          NaN    -7.608472   
CT       -8.405103    -7.948475   -12.411014          NaN    -6.677343   
DC             NaN          NaN          NaN          NaN          NaN   
DE             NaN   -10.274536          NaN          NaN          NaN   
FL      -13.534299   -12.569558   -18.053374          NaN   -10.239592   
GA      -15.018034   -13.858835          NaN          NaN          NaN   
HI       -8.747994    -7.987181          NaN          NaN          NaN   
IA             NaN   -11.726673          NaN          NaN          NaN   
ID      -10.620398   -10.014219          NaN          NaN          NaN   
IL      -10.688482   -10.406940   -16.181427          NaN          NaN   
IN       -8.305096    -7.688026          NaN          NaN    -7.584248   
KS      -10.895518    -9.835536          NaN          NaN    -7.487729   
KY       -8.325745    -8.097982   -14.338234          NaN    -6.942597   
LA       -9.150146    -8.747732          NaN          NaN          NaN   
MA       -7.921997    -7.803463   -13.472194          NaN    -7.741402   
MD       -8.788942    -8.598997          NaN          NaN    -6.491564   
ME      -10.522622    -9.141564          NaN          NaN          NaN   
MI      -12.343430   -11.316613          NaN          NaN          NaN   
MN       -8.661399    -7.883399   -14.481856          NaN          NaN   
MO      -11.183484   -10.206768          NaN          NaN    -9.549612   
MS      -13.744104          NaN          NaN          NaN          NaN   
MT      -10.072184    -9.539963          NaN          NaN          NaN   
NC      -10.721813   -10.172403   -15.234263          NaN          NaN   
ND             NaN          NaN          NaN          NaN          NaN   
NE      -10.225587   -10.011462          NaN          NaN          NaN   
NH      -11.338110   -10.425886          NaN          NaN          NaN   
NJ      -11.063407   -10.389527   -16.379915          NaN    -8.537564   
NM      -11.101464   -11.164200          NaN          NaN          NaN   
NV       -9.527766    -9.015820          NaN          NaN          NaN   
NY       -9.660599    -9.235781   -14.055095          NaN    -6.634022   
OH       -8.540125    -8.629338   -14.069577    -16.63842    -5.551620   
OK      -12.358208   -10.550853          NaN          NaN          NaN   
OR      -11.467104   -11.061180          NaN          NaN          NaN   
PA       -7.378007    -7.672800   -14.663549          NaN    -8.515650   
RI      -10.377028   -12.280813          NaN          NaN          NaN   
SC      -10.695086   -10.028451          NaN          NaN    -9.098535   
SD       -9.568199    -9.218392          NaN          NaN          NaN   
TN      -12.240261   -11.699134          NaN          NaN          NaN   
TX      -11.361813   -10.857851          NaN          NaN          NaN   
UT      -12.066772   -10.625645          NaN          NaN          NaN   
VA       -9.779228    -9.485665          NaN          NaN          NaN   
VT      -11.085710   -10.407638          NaN          NaN          NaN   
WA      -11.374062   -11.446073   -14.825948          NaN          NaN   
WI      -14.299344   -14.884306          NaN          NaN          NaN   
WV      -10.526494    -9.958528          NaN          NaN          NaN   
WY             NaN          NaN          NaN          NaN          NaN   

NDC    2322730      2322830      2322930      2323030      2323130      ...  

#### Exploring the data (very briefly) 
- What is the drug with the highest log-normalized ratio in Hawaii?


In [48]:
medicaid_norm_ndc.loc['HI'].sort_values(ascending=False)

NDC
43386009019    10.511181
43386006019     3.448037
10572010001     2.599661
116200116       2.284969
62175044601     2.195764
                 ...    
76439036590          NaN
99207013070          NaN
99207024005          NaN
99207026012          NaN
99207046330          NaN
Name: HI, Length: 23885, dtype: float64

* Investigate the NDC of the product with the highest log-normalized ratio in Hawaii 
  * What is it used for?

* Compare the value of `Units Reimbursed` that product between HI and other states, (take for instance MA, FL, OR and WA)
* Can you think for reasons why this product has the highest log-normalized ratio in Hawaii?

In [49]:
medicaid_data_w_enrollments_hierarch.loc['HI', 43386009019]

/var/folders/jn/lb25bnzx4js5f2jfpwzh8brw0000gn/T/ipykernel_74618/2593425549.py:1: PerformanceWarning: indexing past lexsort depth may impact performance.
  medicaid_data_w_enrollments_hierarch.loc['HI', 43386009019]


UTILIZATION TYPE PRODUCT NAME  UNITS REIMBURSED  \
STATE NDC                                                           
HI    43386009019             MCOU   GAVILYTE-G          472000.0   
      43386009019             MCOU   GAVILYTE-G          376012.0   
      43386009019             MCOU   GAVILYTE-G       496104006.0   

                   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
STATE NDC                                                             
HI    43386009019                    118.0                  1754.68   
      43386009019                    105.0                  1462.60   
      43386009019                     63.0                   847.56   

                   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
STATE NDC                                                                       
HI    43386009019                     1754.68                            0.00   
      43386009019                     1462.60                            0.00   
      43386009019                      839.61                            7.95   

                               LOCATION  TOTAL MEDICAID ENROLLEES  
STATE NDC                                                          
HI    43386009019  (21.1098, -157.5311)                  340513.0  
      43386009019  (21.1098, -157.5311)                  340513.0  
      43386009019  (21.1098, -157.5311)                  340513.0

In [50]:
medicaid_data_w_enrollments_hierarch.loc['WA', 43386009019]

/var/folders/jn/lb25bnzx4js5f2jfpwzh8brw0000gn/T/ipykernel_74618/1057765572.py:1: PerformanceWarning: indexing past lexsort depth may impact performance.
  medicaid_data_w_enrollments_hierarch.loc['WA', 43386009019]


UTILIZATION TYPE PRODUCT NAME  UNITS REIMBURSED  \
STATE NDC                                                           
WA    43386009019             FFSU   GaviLyte -            58.000   
      43386009019             FFSU   GaviLyte -            65.000   
      43386009019             MCOU   GaviLyte -          2702.118   
      43386009019             MCOU   GaviLyte -          2642.000   
      43386009019             MCOU   GaviLyte -          2753.149   
      43386009019             FFSU   GaviLyte -            87.000   

                   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
STATE NDC                                                             
WA    43386009019                     56.0                  1083.76   
      43386009019                     65.0                  1045.00   
      43386009019                   2677.0                 31709.24   
      43386009019                   2578.0                 28734.28   
      43386009019                   2719.0                 34745.72   
      43386009019                     85.0                  1389.96   

                   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
STATE NDC                                                                       
WA    43386009019                     1083.76                            0.00   
      43386009019                     1031.66                           13.34   
      43386009019                    31691.66                           17.58   
      43386009019                    28732.78                            1.50   
      43386009019                    34743.22                            2.50   
      43386009019                     1387.77                            2.19   

                               LOCATION  TOTAL MEDICAID ENROLLEES  
STATE NDC                                                          
WA    43386009019  (47.3917, -121.5708)                 1771679.0  
      43386009019  (47.3917, -121.5708)                 1771679.0  
      43386009019  (47.3917, -121.5708)                 1771679.0  
      43386009019  (47.3917, -121.5708)                 1771679.0  
      43386009019  (47.3917, -121.5708)                 1771679.0  
      43386009019  (47.3917, -121.5708)                 1771679.0

In [51]:
medicaid_data_w_enrollments_hierarch.loc['MA', 43386009019]

/var/folders/jn/lb25bnzx4js5f2jfpwzh8brw0000gn/T/ipykernel_74618/1093106079.py:1: PerformanceWarning: indexing past lexsort depth may impact performance.
  medicaid_data_w_enrollments_hierarch.loc['MA', 43386009019]


UTILIZATION TYPE PRODUCT NAME  UNITS REIMBURSED  \
STATE NDC                                                           
MA    43386009019             FFSU   GAVILYTE-G             461.0   
      43386009019             MCOU   GAVILYTE-G             486.0   
      43386009019             FFSU   GAVILYTE-G            1086.0   
      43386009019             FFSU   GAVILYTE-G             676.0   
      43386009019             MCOU   GAVILYTE-G             753.0   
      43386009019             MCOU   GAVILYTE-G            1971.0   

                   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
STATE NDC                                                             
MA    43386009019                    294.0                  7135.59   
      43386009019                    471.0                  4587.04   
      43386009019                    479.0                  9185.81   
      43386009019                    499.0                 11456.60   
      43386009019                    727.0                  7264.42   
      43386009019                    925.0                 12118.86   

                   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
STATE NDC                                                                       
MA    43386009019                     5720.40                         1415.19   
      43386009019                     4587.04                            0.00   
      43386009019                     9185.81                            0.00   
      43386009019                     9617.89                         1838.71   
      43386009019                     7264.42                            0.00   
      43386009019                     9820.33                         2298.53   

                              LOCATION  TOTAL MEDICAID ENROLLEES  
STATE NDC                                                         
MA    43386009019  (42.2373, -71.5314)                 1829618.0  
      43386009019  (42.2373, -71.5314)                 1829618.0  
      43386009019  (42.2373, -71.5314)                 1829618.0  
      43386009019  (42.2373, -71.5314)                 1829618.0  
      43386009019  (42.2373, -71.5314)                 1829618.0  
      43386009019  (42.2373, -71.5314)                 1829618.0

In [52]:
medicaid_data_w_enrollments_hierarch.loc['FL', 43386009019]

/var/folders/jn/lb25bnzx4js5f2jfpwzh8brw0000gn/T/ipykernel_74618/2215370880.py:1: PerformanceWarning: indexing past lexsort depth may impact performance.
  medicaid_data_w_enrollments_hierarch.loc['FL', 43386009019]


UTILIZATION TYPE PRODUCT NAME  UNITS REIMBURSED  \
STATE NDC                                                           
FL    43386009019             MCOU   GAVILYTE-G          596000.0   
      43386009019             MCOU   GAVILYTE-G          676000.0   
      43386009019             MCOU   GAVILYTE-G          576000.0   

                   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
STATE NDC                                                             
FL    43386009019                    149.0                  1910.49   
      43386009019                    169.0                  2137.42   
      43386009019                    144.0                  1920.29   

                   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
STATE NDC                                                                       
FL    43386009019                     1906.79                             3.7   
      43386009019                     2137.42                             0.0   
      43386009019                     1920.29                             0.0   

                             LOCATION  TOTAL MEDICAID ENROLLEES  
STATE NDC                                                        
FL    43386009019  (27.8333, -81.717)                 3808334.0  
      43386009019  (27.8333, -81.717)                 3808334.0  
      43386009019  (27.8333, -81.717)                 3808334.0

In [53]:
medicaid_data_w_enrollments_hierarch.loc['AZ', 43386009019]

/var/folders/jn/lb25bnzx4js5f2jfpwzh8brw0000gn/T/ipykernel_74618/2466570507.py:1: PerformanceWarning: indexing past lexsort depth may impact performance.
  medicaid_data_w_enrollments_hierarch.loc['AZ', 43386009019]


UTILIZATION TYPE PRODUCT NAME  UNITS REIMBURSED  \
STATE NDC                                                           
AZ    43386009019             MCOU   GAVILYTE-G         7828400.0   
      43386009019             MCOU   GAVILYTE-G         6060000.0   
      43386009019             MCOU   GAVILYTE-G         7040000.0   

                   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
STATE NDC                                                             
AZ    43386009019                   1909.0                 28182.24   
      43386009019                   1499.0                 20287.85   
      43386009019                   1731.0                 26035.13   

                   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
STATE NDC                                                                       
AZ    43386009019                    28131.21                           51.03   
      43386009019                    20267.83                           20.02   
      43386009019                    25939.95                           95.18   

                               LOCATION  TOTAL MEDICAID ENROLLEES  
STATE NDC                                                          
AZ    43386009019  (33.7712, -111.3877)                 1740520.0  
      43386009019  (33.7712, -111.3877)                 1740520.0  
      43386009019  (33.7712, -111.3877)                 1740520.0

* Find and list all unique `NDC`s for which the difference between the largest and second large log-normalized ratio by state is at least 10.

* For instance:
  * The highest log-normalized ratio for `00591289749` (`AZACITIDIN`) is OK where it has a log-normalized ratio  of `-1.025642`.
  * The second highest log-normalized ratio for `00591289749` is in `GA` where is has a log-normalized ration of  `-12.623428`
  


In [60]:
medicaid_norm_ndc

NDC    2143380      2143480      2144511      2144527      2197590      \
STATE                                                                    
AK       -9.609109   -10.008280          NaN          NaN          NaN   
AL       -9.940595    -9.805485          NaN          NaN          NaN   
AR      -12.985157   -12.657103          NaN          NaN          NaN   
AZ      -11.533870   -10.821194          NaN          NaN          NaN   
CA      -10.926863   -10.698372   -16.127018          NaN    -9.464364   
CO      -10.862982    -9.925954          NaN          NaN    -7.608472   
CT       -8.405103    -7.948475   -12.411014          NaN    -6.677343   
DC             NaN          NaN          NaN          NaN          NaN   
DE             NaN   -10.274536          NaN          NaN          NaN   
FL      -13.534299   -12.569558   -18.053374          NaN   -10.239592   
GA      -15.018034   -13.858835          NaN          NaN          NaN   
HI       -8.747994    -7.987181          NaN          NaN          NaN   
IA             NaN   -11.726673          NaN          NaN          NaN   
ID      -10.620398   -10.014219          NaN          NaN          NaN   
IL      -10.688482   -10.406940   -16.181427          NaN          NaN   
IN       -8.305096    -7.688026          NaN          NaN    -7.584248   
KS      -10.895518    -9.835536          NaN          NaN    -7.487729   
KY       -8.325745    -8.097982   -14.338234          NaN    -6.942597   
LA       -9.150146    -8.747732          NaN          NaN          NaN   
MA       -7.921997    -7.803463   -13.472194          NaN    -7.741402   
MD       -8.788942    -8.598997          NaN          NaN    -6.491564   
ME      -10.522622    -9.141564          NaN          NaN          NaN   
MI      -12.343430   -11.316613          NaN          NaN          NaN   
MN       -8.661399    -7.883399   -14.481856          NaN          NaN   
MO      -11.183484   -10.206768          NaN          NaN    -9.549612   
MS      -13.744104          NaN          NaN          NaN          NaN   
MT      -10.072184    -9.539963          NaN          NaN          NaN   
NC      -10.721813   -10.172403   -15.234263          NaN          NaN   
ND             NaN          NaN          NaN          NaN          NaN   
NE      -10.225587   -10.011462          NaN          NaN          NaN   
NH      -11.338110   -10.425886          NaN          NaN          NaN   
NJ      -11.063407   -10.389527   -16.379915          NaN    -8.537564   
NM      -11.101464   -11.164200          NaN          NaN          NaN   
NV       -9.527766    -9.015820          NaN          NaN          NaN   
NY       -9.660599    -9.235781   -14.055095          NaN    -6.634022   
OH       -8.540125    -8.629338   -14.069577    -16.63842    -5.551620   
OK      -12.358208   -10.550853          NaN          NaN          NaN   
OR      -11.467104   -11.061180          NaN          NaN          NaN   
PA       -7.378007    -7.672800   -14.663549          NaN    -8.515650   
RI      -10.377028   -12.280813          NaN          NaN          NaN   
SC      -10.695086   -10.028451          NaN          NaN    -9.098535   
SD       -9.568199    -9.218392          NaN          NaN          NaN   
TN      -12.240261   -11.699134          NaN          NaN          NaN   
TX      -11.361813   -10.857851          NaN          NaN          NaN   
UT      -12.066772   -10.625645          NaN          NaN          NaN   
VA       -9.779228    -9.485665          NaN          NaN          NaN   
VT      -11.085710   -10.407638          NaN          NaN          NaN   
WA      -11.374062   -11.446073   -14.825948          NaN          NaN   
WI      -14.299344   -14.884306          NaN          NaN          NaN   
WV      -10.526494    -9.958528          NaN          NaN          NaN   
WY             NaN          NaN          NaN          NaN          NaN   

NDC    2322730      2322830      2322930      2323030      2323130      ...  

In [58]:
## WRITE YOUR CODE HERE 
medicaid_norm_ndc.nlargest(2, medicaid_norm_ndc.columns)

NDC    2143380      2143480      2144511      2144527      2197590      \
STATE                                                                    
PA       -7.378007    -7.672800   -14.663549          NaN    -8.515650   
MA       -7.921997    -7.803463   -13.472194          NaN    -7.741402   

NDC    2322730      2322830      2322930      2323030      2323130      ...  \
STATE                                                                   ...   
PA       -4.623149    -3.091035    -2.718405          NaN          NaN  ...   
MA       -5.414724    -4.592154    -4.205626          NaN          NaN  ...   

NDC    76439035930  76439035990  76439036290  76439036390  76439036490  \
STATE                                                                    
PA             NaN          NaN          NaN          NaN    -9.698115   
MA             NaN          NaN          NaN          NaN          NaN   

NDC    76439036590  99207013070  99207024005  99207026012  99207046330  
STATE                                                                   
PA       -9.289037          NaN          NaN          NaN          NaN  
MA             NaN          NaN          NaN          NaN          NaN  

[2 rows x 23885 columns]

- The Drug `AZACITIDINE` has a very high normalized UNITS REIMBURSED in OK compared to other states.
   - Normalized log value is -1.025642 (or a ratio 0.49119167009735065)
   - Second highest state has a log value of -12.623428 (0.000158478197834722)
- Oklahoma is not a high-incidence state for cancer
- Could the following explain what is happening in Oklahoma?

https://www.centerwatch.com/clinical-trials/listings/92093/acute-myeloid-leukemia-aml-study-asp2215-gilteritinib-by/?&geo_lat=35.4675602&geo_lng=-97.5164276&radius=10